address-space-controller/src/main/java/io/enmasse/controller/CreateController.java

```diff
@@ -130,10 +132,25 @@ public AddressSpace handle(AddressSpace addressSpace) throws Exception {
                 kubernetes.apply(resourceList);
                 eventLogger.log(AddressSpaceUpgraded, "Upgraded address space", Normal, ControllerKind.AddressSpace, addressSpace.getName());
                 addressSpace.putAnnotation(AnnotationKeys.APPLIED_INFRA_CONFIG, mapper.writeValueAsString(desiredInfraConfig));
+                addressSpace.putAnnotation(AnnotationKeys.APPLIED_PLAN, addressSpace.getPlan());
             } else {
                 log.info("Version of desired config ({}) does not match controller version ({}), skipping upgrade", desiredInfraConfig.getVersion(), version);
             }
+        } else if (!addressSpace.getPlan().equals(addressSpace.getAnnotation(AnnotationKeys.APPLIED_PLAN))) {
+            kubernetes.ensureServiceAccountExists(addressSpace);
+            KubernetesList resourceList = new KubernetesListBuilder()
+                    .addAllToItems(infraResourceFactory.createInfraResources(addressSpace, desiredInfraConfig))
+                    .build();
+            addAppliedInfraConfigAnnotation(resourceList, desiredInfraConfig);
+
+            log.info("Updating address space plan {}", addressSpace);
+
+            kubernetes.apply(resourceList);
+            eventLogger.log(AddressSpaceChanged, "Changed address space plan", Normal, ControllerKind.AddressSpace, addressSpace.getName());
+            addressSpace.putAnnotation(AnnotationKeys.APPLIED_INFRA_CONFIG, mapper.writeValueAsString(desiredInfraConfig));
+            addressSpace.putAnnotation(AnnotationKeys.APPLIED_PLAN, addressSpace.getPlan());
         }
```

Comment:
No, I don't believe its possible to distinguish them since it expresses the desired state. I don't think we should retain an audit log of changes on the resource itself. However, we do emit the AddressSpaceChanged event, which may create a k8s event giving you the audit log that way.

On the ongoing part, changing the address space plan will cause the standard controller to be redeployed (no effect on brokered address space agent since we don't have more than 1 address space plan) to pick up the change. We have a status check checking the deployments of the address space to become ready, so if applying the change runs into any errors, the address space health alert should fire.

Reply:
The fact that we currently only provide one brokered address space plan is just a function of our "example" plans though, is it not?  Theoretically we could have different broker plans with (for instance) difference RAM or PV size requirements?  What would happen then?  On the standard address space, what happens if the new address space plan sets a limit on the number of routers/brokers lower than that required by the cumulative requirements of the addresses defined within the space?

Reply:
> The fact that we currently only provide one brokered address space plan is just a function of our "example" plans though, is it not? Theoretically we could have different broker plans with (for instance) difference RAM or PV size requirements? What would happen then?

To do that, the new plan would refer to a different infra config/the infra config would be changed. This will in turn trigger the ' infra upgrade/change' code path and will cause relevant components to get redeployed.

> On the standard address space, what happens if the new address space plan sets a limit on the number of routers/brokers lower than that required by the cumulative requirements of the addresses defined within the space?

No such check at present, but I agree we should add one. This check should be added to the address space controller before applying the new change. In this case, i think it should not apply the new plan and provide a status message saying that plan change cannot be applied.

Reply:
> > The fact that we currently only provide one brokered address space plan is just a function of our "example" plans though, is it not? Theoretically we could have different broker plans with (for instance) difference RAM or PV size requirements? What would happen then?
> 
> To do that, the new plan would refer to a different infra config/the infra config would be changed. This will in turn trigger the ' infra upgrade/change' code path and will cause relevant components to get redeployed.

Which, if the PV size requirement changed, would magically erase all their messages?  Would a change to the RAM requirements retain the same PV?  I'd want to avoid people being shocked by all their messages suddenly disappearing.

> > On the standard address space, what happens if the new address space plan sets a limit on the number of routers/brokers lower than that required by the cumulative requirements of the addresses defined within the space?
> 
> No such check at present, but I agree we should add one. This check should be added to the address space controller before applying the new change. In this case, i think it should not apply the new plan and provide a status message saying that plan change cannot be applied.

In this case what would happen to the annotations on the address space that are updated by the above code?

Reply:
> Which, if the PV size requirement changed, would magically erase all their messages? Would a change to the RAM requirements retain the same PV? I'd want to avoid people being shocked by all their messages suddenly disappearing.

Changing PV size is not supported by Kubernetes/OpenShift at present, so changing storage size will not affect existing brokers (this applies to upgrade/config change in general).

> In this case what would happen to the annotations on the address space that are updated by the above code?

The annotation should only get set once the plan has passed the quota check. This means the operation will be retried until set of addresses has been reduced below the quota to be applied.

Reply:
> Changing PV size is not supported by Kubernetes/OpenShift at present, so changing storage size will not affect existing brokers (this applies to upgrade/config change in general).

OK - so though the plan (and infra) will look as though they have been updated to the new plan, in fact some aspects will not have changed... probably worth an FAQ/note somewhere for those not familiar with this aspect of Kubernetes...  Presumably this means that if you have a brokered space which has an overlarge journal, you are just out of luck?

Reply:
Agreed we should add a note on this to the docs. It looks like increasing PV size is supported in 1.11/3.11, but it will depend on the underlying storage provisioner. See https://kubernetes.io/blog/2018/07/12/resizing-persistent-volumes-using-kubernetes/ for more info.

However, our code filters out changes to the PVC right now, but this could be changed via some parameter perhaps (allowResize) or something like that.

Reply:
Yeah - if we can support increasing the size in the PVC that would potentially get people out of a "running out of space, don't want to lose messages" scenario.

Is it worth writing some tooling to detect deployed infra where the actual doesn't meet the expected for the plan (for instance they've requested a new plan and the storage size hasn't actually changed)?  Are there other parts of the infra that might similarly not update?